packages/oui-navbar/src/navbar.controller.js

```diff
@@ -1,52 +1,93 @@
         "ngInject";
 
         this.$attrs = $attrs;
         this.$document = $document;
         this.$element = $element;
         this.$timeout = $timeout;
         this.config = ouiNavbarConfiguration;
-        this.navbarService = NavbarService;
+
+        this.keyboardNav = {};
         this.KEYBOARD_KEYS = KEYBOARD_KEYS;
     }
 
+    addItemToGroup (item, groupName) {
+        // Create the group if it doesn't exist
+        if (angular.isUndefined(this.keyboardNav[groupName])) {
+            this.keyboardNav[groupName] = [];
+        }
+
+        // Add item to the group
+        this.keyboardNav[groupName].push(item);
+    }
+
+    getGroup (groupName) {
+        return this.keyboardNav[groupName];
+    }
+
+    // Set focus to an item of a group
+    setFocusTo (groupName, index) {
+        // Add a delay to force focus
+        const delay = 50;
+        this.$timeout(() => {
+            if (get(this.keyboardNav, `${groupName}[${index}]`)) {
+                this.keyboardNav[groupName][index].focus();
+            }
+        }, delay);
+    }
+
     toggleMenu (state, isInternalNav) {
-        // Update navbar navigation
-        this.navigation = this.navbarService.toggleMenu(state, isInternalNav);
+        if (state) {
+            // Reset navigation if not internal navigation
+            if (!isInternalNav && (!this.navigation || !this.navigation[state])) {
+                this.navigation = {};
+            }
+
+            if (isInternalNav || !this.navigation[state]) {
+                // Toggle menu if internal navigation or state is undefined
+                this.navigation[state] = !this.navigation[state];
+            } else if (this.navigation[state]) {
+                // Else close all menus
+                this.navigation = null;
+            }
+
+            // Focus first list item when opened
+            if (this.navigation && this.navigation[state]) {
+                // Add a little delay to avoid transition bug on Webkit
+                this.setFocusTo(state, 0);
+            }
+        } else {
+            // Close all menus
+            this.navigation = null;
+        }
     }
 
-    updateLinks () {
+    $onInit () {
         // If no togglerLinks attribute, we use the value of mainLinks
         if (!angular.isDefined(this.$attrs.togglerLinks) && angular.isDefined(this.$attrs.mainLinks)) {
             this.togglerLinks = this.mainLinks;
-            this.togglerLoaded = true;
         }
-    }
-
-    $onInit () {
-        this.updateLinks();
 
         // Support presence of attribute 'fixed'
         addBooleanParameter(this, "fixed");
     }
 
-    $onChanges (changes) {
-        this.updateLinks();
-
-        // Get togglerLinks changes for the loader
-        if (changes.togglerLinks) {
-            this.togglerLoaded = !!changes.togglerLinks.currentValue;
-        }
+    $onDestroy () {
+        this.$document
```

Comment:
We've got listeners on the whole document ? What are they for ?

Reply:
keydown for ESC and click to close opened menu :)